### ISSUE TITLE
chararray subscription returns empty string instead of empty list (migrated from Trac #1948)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3496]
Original ticket http://projects.scipy.org/numpy/ticket/1948
Reported 2011-09-06 by trac user jazzer, assigned to unknown.

Current behaviour (1.5.1, 1.6.1):
    >>> import numpy
    >>> a=numpy.rec.array([("a",1),("bb",2),("cc",3)],dtype=[('s','|S4'),('n','i4')])
    >>> print a.s[a.s>'b']
    ['bb' 'cc']
    >>> print a.s[a.s>'c']
    ['cc']
    >>> print a.s[a.s>'d']

```
>>> a.s[a.s>'d']
''
```

The correct behaviour (1.2.1):
    >>> print a.s[a.s>'d']
    []
    >>> a.s[a.s>'d']
    chararray([],
          dtype='|S4')

The bug was introduced in the commit:7deb4d5db4282e1e1c86, file numpy/core/defchararray.py (I'm not sure to which NumPy release this commit corresponds, it's somewhere between 1.2.1 and 1.5.1):
         def **getitem**(self, obj):
             val = ndarray.**getitem**(self, obj)
    -        if isinstance(val, (string_, unicode_)):
    +        if issubclass(val.dtype.type, character):
                 temp = val.rstrip()
Reverting the change in this line fixes the bug.

**Comment by thouis** [comment]
Attachment in Trac by trac user jazzer, 2011-09-06: [chararray-empty-string.patch](http://projects.scipy.org/numpy/attachment/ticket/1948/chararray-empty-string.patch)

**Comment by thouis** [comment]
Comment in Trac by trac user jazzer, 2011-09-06

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-02-20

Empty list seems correct to me, but there may have been a reason for this change.  Needs review for 1.7.0.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-03-25

The proposed fix (returning []) is consistent with how plain arrays with string dtypes work:
    >>> >>> s = np.array(['a', 'bb', 'cc'], dtype='S4')
    >>> >>> s[[]]
    array([], 
          dtype='|S4')
    >>> 
    >>> >>> a = np.rec.array([("a",1),("bb",2),("cc",3)],dtype=[('s','|S4'),('n','i4')])
    >>> >>> a.s[[]]
    []
    >>> 
    >>> >>> s == a.s
    array([ True,  True,  True], dtype=bool)
    >>> 
    >>> >>> type(s)
    <type 'numpy.ndarray'>
    >>> >>> type(a.s)
    <class 'numpy.core.defchararray.chararray'>

The proposed fix does break another test though:
    ======================================================================
    FAIL: test_expandtabs (test_defchararray.TestMethods)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/Users/rgommers/Code/numpy/numpy/core/tests/test_defchararray.py", line 330, in test_expandtabs
        assert_(T[2][0] == asbytes('123      345'))
      File "/Users/rgommers/Code/numpy/numpy/testing/utils.py", line 34, in assert_
        raise AssertionError(msg)
    AssertionError

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-03-25

Related:
    >>> np.chararray([])
    chararray('', 
          dtype='|S1')
    >>> np.chararray([]).size
    1

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-03-25

```
>>> np.chararray([]).size
1
>>> np.chararray([]).shape
()
>>> np.array([])
array([], dtype=float64)
>>> np.array([]).size
0
>>> np.array([]).shape
(0,)
```

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-03-25

 A fix that doesn't break anything is to replace
    if issubclass(val.dtype.type, character):
by
    if issubclass(val.dtype.type, character) and not len(val) == 0:

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-04-07

I sent a PR with my proposed fix: https://github.com/numpy/numpy/pull/247

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-04-08

Fixed in commit:b6b5ba1
